[AAP-45287] Fix IPv6 handling for Redis

jshimkus-rh opened the pull request:
## Description
<!-- Mandatory: Provide a clear, concise description of the changes and their purpose -->
- What is being changed?
* Extending the class of addresses supported for Redis clients to include IPv6 values.

- Why is this change needed?
* Current Redis client rejects IPv6 address as invalid.

- How does this change address the issue?
Accepts IPv6 addresses in the common forms of <addr>:<port> and [<addr>]:<port>.

## Type of Change
<!-- Mandatory: Check one or more boxes that apply -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Test update
- [ ] Refactoring (no functional changes)
- [ ] Development environment change
- [ ] Configuration change

## Self-Review Checklist
<!-- These items help ensure quality - they complement our automated CI checks -->
- [x] I have performed a self-review of my code
- [x] I have added relevant comments to complex code sections
- [ ] I have updated documentation where needed
- [ ] I have considered the security impact of these changes
- [ ] I have considered performance implications
- [x] I have thought about error handling and edge cases
- [x] I have tested the changes in my local environment

## Testing Instructions
### Steps to Test
1. Deploy AAP environment in an IPv6-only environment *not utilizing host names for Redis client.  Installer operates as such.

### Expected Results
Redis client connections successfully instantiated.

jshimkus committed:
[AAP-45287] Fix IPv6 handling for Redis

jshimkus-rh committed:
Merge branch 'devel' into devel

bhavenst commented on test_app/tests/lib/utils/test_address.py:
```diff
@@ -0,0 +1,99 @@
+        ("::1", AddressType.IPv6),
+        ("[::1]", AddressType.HOSTNAME),
+        ("2600:1f18:218b:5902:e5d4:54de:fdc1:24b8", AddressType.IPv6),
+        ("[2600:1f18:218b:5902:e5d4:54de:fdc1:24b8]", AddressType.HOSTNAME),
```
Aren't these bracket notation ipv6 addresses?  I thought the [] was just to avoid interpreting the port as part of the address.  Probably wrong, but it seems odd that these resolve to hostname.

bhavenst replied:
Oh, noticed the note stating that.  Interesting, is that a general thing or just for purposes of this usage?

jshimkus-rh replied:
It's to differentiate between a raw address that must be wrapped in []s and one already so wrapped.  I chose to use hostname as that's the analogous common use case we have: `https://[<ipv6>]:<port>/....`

bhavenst commented on test_app/tests/lib/utils/test_address.py:
```diff
@@ -0,0 +1,99 @@
+    assert is_hostname_address_string(address)
+
+
+@pytest.mark.parametrize(
```
Maybe a couple of invalid cases? 127.0.0.257 for instance.

bhavenst commented on ansible_base/lib/redis/client.py:
```diff
@@ -160,13 +161,15 @@ def _get_hosts(self) -> None:
+                    had_host_errors = True
+                    continue
+                if not port_string:
                     logger.error(f"Specified cluster_host {host_port} is not valid; it needs to be in the format <host>:<port>")
```
Need to expand this "needs to be in format ..." list?

jshimkus committed:
Address black format complaints.

jshimkus committed:
Add some additional test cases.

jshimkus committed:
Expound on required address type in log message.

jshimkus committed:
Fix a black restructuring that caused a sonarqube complaint.

tznamena commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,109 @@
+import re
+
+
+class AddressType(enum.Enum):
```
Some documentation would be nice for this class.

jshimkus-rh replied:
Added.

tznamena replied:
Awesome docs! :+1:

jshimkus committed:
Add doc string to AddressType.

john-westcott-iv commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,125 @@
+        # Only an address type of IPv6 is considered valid here.
+        # We don't want to treat any other specification as valid.
+        if address_type == AddressType.IPv6:
+            return AddressType.HOSTNAME
```
Is HOSTNAME correct here? Wouldn't it be an IPv6 or do we consider [<ip v6 address>] as a hostname?

jshimkus-rh replied:
The comment before the preceding if can be more explicit.  The intention here is two-fold:
1. minimize the type space; i.e., not have two kinds of IPv6
2. minimize the logic that a client has to implement 

In the case of 2 the way in which a []ed IPv6 is used is as if it were a hostname.  That is, no special processing is required (beyond convenience this is an additional reason there's a method that does the splitting into "address" and "port"; otherwise the client would still have to special case processing). 

So the only, if it matters to the component, address type it may have to care about is the single IPv6 type.

jshimkus-rh replied:
Changed to categorize IPv6, wrapped or not, as IPv6.  Added a property to the AddressTypeResponse which will return the bracketed IPv6 or None if the address type is not IPv6.

john-westcott-iv commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,125 @@
+    return (AddressType.UNKNOWN, "", "")
+
+
+def is_hostname_address_string(address_string: str) -> bool:
```
This name of this one feels weird to me. Can we call it `is_hostname` or would that conflict with something else?

john-westcott-iv replied:
Also, do we want an `is_hostname_with_port` method?

relrod replied:
I'd go further and say, can we drop the `_string` everywhere? Both in the function names, and also in the function parameter? In both cases, the type of the input tells us it's a string.

I'd prefer e.g. `classify_address(address: str)`, etc.

jshimkus-rh replied:
> This name of this one feels weird to me. Can we call it `is_hostname` or would that conflict with something else?

@john-westcott-iv Sure, we can shorten the name.  In some sense I'd prefer that based on my perspective that python code should be using (at the least) the terminal namespace to access offerings; primarily for some identification as to origin but also to avoid name collisions.

jshimkus-rh replied:
> Also, do we want an `is_hostname_with_port` method?

Possibly.  I was thinking about how far we want to split the identification breakdown.

I included the is_* methods mostly because that's what I did in the gateway code.  But nothing uses them and we if only provide the full breakdown (type, address, port) method we eliminate the inconsistencies in which is_ipv6 with <addr>:<port> reports "No" but the full breakdown reports IPv6.

So I'm going to eliminate everything other than the full breakdown.

john-westcott-iv commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,125 @@
+        address_string and port_string are only valid (though port string may
+        be an empty string) when AddressType is not AddressType.UNKNOWN.
+    """
+    address_type = classify_address_string(address_string)
```
When I'm looking at this, I'm wondering if we can shrink this logic with a regex like:
```
address_re = re.compile('^\[?([a-zA-Z0-9:]*)\]?(:([^:]+))?')
address_re.matches(address_string)
```
Its not quite working, `localhost:` passes the check. But it feels like something like this might be able to reduce some logic in here.

jshimkus-rh replied:
Added this discussion of the regex.
```
    # The regex may appear more complicated than it actually is. First you can
    # ignore the "?:" which simply makes the group in which it appears
    # non-capturing.
    #
    # Second you can conceptually collapse the portions delineated by "{" and
    # "}" to "*". The bracked portions simply put a limit on minimum and
    # maximum lengths of the preceding construct.
    #
    # With the two above changes you'll see that there's effectively only one
    # construct in the regex:
    #
    #   [a-zA-Z]([a-zA-Z0-9-]*[a-zA-Z0-9])?
    #
    # It's second usage simply includes the requirement for a leading "." and
    # allows it to be specified zero or more times.
```

relrod commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,125 @@
+        pass
+
+    # Basic hostname check (can be expanded for more rigorous validation)
+    if re.match(r"^[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$", address_string):
```
Can you leave a comment with where this regex came from, assuming you didn't write it yourself? Or alternatively, maybe split it into multiple lines and comment what is going on here?

This is a moderately complex regex, I worry about being able to maintain it going forward.

jshimkus-rh replied:
Good idea re: the comment.  
I tried the splitting originally but our CI linting complained about it.

< aside> The AI generated version of the core categorization provided a simpler regex that just didn't work (not even for names like "a-hostname".  But asking solely for a regex that would identify hostnames gave the one in the code. < /aside>

relrod commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,125 @@
+    if address_type != AddressType.UNKNOWN:
+        return address_type
+
+    # We could be dealing with an IPv6 address wrapped in [].
```
Rather than having to parse it twice in this case, would it be better to just strip the `[` and `]` at the beginning, parse the string, then if we get IPv6 back and the original string had `[` and `]` then we characterize it as `HOSTNAME`?

In other words, something like this (untested):


```python
is_bracketed = address_string[0] == '[' and address_string[-1] == ']'
address_type = _classify_address_string(address_string.strip('[]'))
if address_type == AddressType.IPv6 and is_bracketed:
    return AddressType.HOSTNAME
return address_type
```

And probably a check above to make sure we're not given the empty string as input, otherwise we'd crash on indexing into it.

jshimkus-rh replied:
Not unreasonable re: parsing.

Definitely yes re: empty string.

relrod commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,125 @@
+    return AddressType.UNKNOWN
+
+
+def classify_address_string(address_string) -> AddressType:
```
```suggestion
def classify_address_string(address_string: str) -> AddressType:
```

relrod commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,125 @@
+    UNKNOWN = "unknown"
+
+
+def _classify_address_string(address_string) -> AddressType:
```
```suggestion
def _classify_address_string(address_string: str) -> AddressType:
```

relrod commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,125 @@
+    return AddressType.UNKNOWN
+
+
+def classify_and_split_address_string(address_string) -> tuple[AddressType, str, str]:
```
```suggestion
def classify_and_split_address_string(address_string: str) -> tuple[AddressType, str, str]:
```

relrod commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,125 @@
+    address_type = classify_address_string(address_string)
+    if address_type != AddressType.UNKNOWN:
+        # A known type with no port.
+        return (address_type, address_string, "")
```
Is there a good reason for port to be a string, beyond "might be empty"?

If not, I'd prefer `Optional[int]` here.

jshimkus-rh replied:
It minimizes modifying existing code which already deals with the port as split from the string.

I'm not saying it's a wrong idea but minimizing changes to existing code being able to use this functionality was the reason.

relrod commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,125 @@
+        return (address_type, address_string, port)
+
+    # An unknown address type.
+    return (AddressType.UNKNOWN, "", "")
```
I'd consider making the address part an `Optional[str]` and have it be `None` here rather than empty string.

jshimkus-rh replied:
Considered that as I'm not a fan of non-explicit comparison of a value as boolean (unless it *is* a boolean).  In this case "" and None would be logically equivalent.

Also, see comment about minimizing changes to existing code.

relrod commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,125 @@
+    return classify_address_string(address_string) == AddressType.IPv4
+
+
+def is_ipv6_address_string(address_string: str) -> bool:
```
Not a blocker, but: Future optimization for these wrapper functions would be to _only_ check the thing we care about in the cases where we can.

In other words, as it stands right now, for `is_ipv6_address_string()` to work, it will always try to be parsed as v4 first, and we should be able to avoid that.

jshimkus-rh replied:
Gonna change to just the one categorization method so this won't be applicable any more.

jshimkus committed:
Address review comments.

Reduces the functionality to one method as a single provided
capability decomposing a potential address into all relevant
parts for consistency.

PabloHiro commented on test_app/tests/lib/utils/test_address.py:
```diff
@@ -0,0 +1,32 @@
+        ("[::1]:1234", AddressTypeResponse(AddressType.HOSTNAME, "[::1]", "1234")),
+        ("2600:1f18:218b:5902:e5d4:54de:fdc1:24b8", AddressTypeResponse(AddressType.IPv6, "2600:1f18:218b:5902:e5d4:54de:fdc1:24b8")),
+        ("[2600:1f18:218b:5902:e5d4:54de:fdc1:24b8]", AddressTypeResponse(AddressType.HOSTNAME, "[2600:1f18:218b:5902:e5d4:54de:fdc1:24b8]")),
+        ("2600:1f18:218b:5902:e5d4:54de:fdc1:24b8:1234", AddressTypeResponse(AddressType.IPv6, "2600:1f18:218b:5902:e5d4:54de:fdc1:24b8", "1234")),
```
This should throw an error imho. No unbracketed IPv6+Port combination should be fed to Gateway in the first place. And if it is, throw an error instead of trying to accommodate for that.

jshimkus-rh replied:
As discussed with @PabloHiro in the IPv6 WG channel unless/until the current mixture of IPv6 specification is standardized and implemented we cannot impose a standard here without potentially breaking deployments.

PabloHiro commented on test_app/tests/lib/redis/test_client.py:
```diff
@@ -209,6 +209,13 @@ def test_redis_cluster_mget_raises_expected_exception():
+        ('[::1]:1', False, 1),
+        ('2600:1f18:218b:5902:e5d4:54de:fdc1:24b8:1', False, 1),
+        ('[2600:1f18:218b:5902:e5d4:54de:fdc1:24b8]:1', False, 1),
+        ('[::1]:1,2600:1f18:218b:5902:e5d4:54de:fdc1:24b8:1', False, 2),
```
Lines 213 and 215 should throw errors too

tznamena replied:
Agree with previous comment(address must be between `[` and `]` when followed by port number) and I would suggest to use a different number for port number to make it more obvious. Since segments of the address are hex, using a five digit number for port number would make it very obvious.

jshimkus-rh replied:
As discussed with @PabloHiro in the IPv6 WG channel unless/until the current mixture of IPv6 specification is standardized and implemented we cannot impose a standard here without potentially breaking deployments.

jshimkus committed:
Add discussion of hostname regex.

tznamena commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,156 @@
+    port: str = ""
+
+
+def _classify_base_address(address: str) -> AddressType:
```
There are two methods here with very similar name, same parameters, same return type and very similar doc(1 sentence difference).
Can we rename these or consolidate them?

jshimkus-rh replied:
They are functionally different in that one (the base version) classifies whatever it is passed.  The non-base version adds the special handling of []-enclosed IPv6.
We can't reduce to one and simply recurse lest we end up interpreting something like this, `[[[[[[[[<ipv6>]:<port>]]]]]]]`, as bracketed IPv6.
We'd have to duplicate code.

tznamena commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,156 @@
+    """
+    Categorizes a given string as IPv4, IPv6, hostname, or unknown.
+
+    An IPv6 address wrapped in [] is considered a hostname.
```
Does it make sense to support address wrapped in `[` and `]` but not have a port following it?

jshimkus-rh replied:
Yes.  E.g., `[::1]` as the IPv6 localhost to use in a URL or any address utilizing the default HTTP port.

jshimkus committed:
Classify IPv6, wrapped or not, as IPv6.

Add a property to AdressTypeResponse which returns the bracketed IPv6 address
or None if the address type is not IPv6.

john-westcott-iv commented on ansible_base/lib/redis/client.py:
```diff
@@ -160,22 +161,27 @@ def _get_hosts(self) -> None:
+                if response.type == address.AddressType.UNKNOWN:
+                    logger.error(
+                        f"Specified cluster_host {host_port} is not valid; "
+                        "it is of an unknown address type and must be one of <hostname>:<port>, <ipv4>:<port> or <ipv6>:<port>"
```
Would this be more accurate?
```suggestion
                        "it is of an unknown address type and must be one of <hostname>:<port>, <ipv4>:<port> or [<ipv6>]:<port>"
```

jshimkus-rh replied:
It's probably better that way.  I considered adding both IPv6 but that seemed extreme.

john-westcott-iv commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,159 @@
+    """
+
+    type: AddressType
+    address: str = ""
+    port: str = ""
```
Non-blocking just curious why empty string instead of None?

jshimkus-rh replied:
Originally (when returning a tuple) to try to minimize necessary client code changes to just a call to the functionality.  Now, None probably makes more sense.

john-westcott-iv commented on ansible_base/lib/utils/address.py:
```diff
@@ -0,0 +1,159 @@
+        # We don't want to treat any other specification as valid.
+        if response.type == AddressType.IPv6:
+            return response
+        return AddressTypeResponse(AddressType.UNKNOWN, address)
```
Can we add a comment in here? Or, alternatively, should we assume its a hostname if its not an IPv6 address? Would we care if someone did `[<hostname>]` or `[<ip v4 address]`?
```suggestion
        # Returning unknown because we gor a [<address>] type string but the address inside wasn't IPv6 but nothing else should be encapsulated with []
        return AddressTypeResponse(AddressType.UNKNOWN, address)
```

jshimkus-rh replied:
Adding a comment is fine.  We want to reject anything that should not be wrapped in brackets.  I'll add the comment separately 'cause of the misspelling. 😄

john-westcott-iv reviewed (approved):
Approving with minor comments/questions.

jshimkus-rh committed:
Update ansible_base/lib/redis/client.py

Specify the expected IPv6 format to include brackets.

Co-authored-by: John Westcott IV <32551173+john-westcott-iv@users.noreply.github.com>

jshimkus committed:
Make AddressResponseType port value an optional.

Add comment that the only valid type of address that can be wrapped in []s is IPv6.
